rust/src/array.rs

```diff
@@ -210,6 +210,48 @@ macro_rules! def_primitive_array {
                 n
             }
 
+            pub fn add(&self, other: &PrimitiveArray<$native_ty>) -> PrimitiveArray<$native_ty> {
+                self.math_helper(other, |a, b| a + b)
+            }
+
+            pub fn subtract(
+                &self,
+                other: &PrimitiveArray<$native_ty>,
+            ) -> PrimitiveArray<$native_ty> {
+                self.math_helper(other, |a, b| a - b)
+            }
+
+            pub fn multiply(
+                &self,
+                other: &PrimitiveArray<$native_ty>,
+            ) -> PrimitiveArray<$native_ty> {
+                self.math_helper(other, |a, b| a * b)
+            }
+
+            pub fn divide(&self, other: &PrimitiveArray<$native_ty>) -> PrimitiveArray<$native_ty> {
+                self.math_helper(other, |a, b| a / b)
```

Comment:
Should we check if `b` is zero and push a null if it is?

Reply:
I believe divide by zero should be an error, as it would be usually

Reply:
However, we don't want a panic ... so probably divide should return a Result<> and should check first for any 0 values before attempting the operation

Reply:
Yea, maybe returning a `Result` is more appropriate.

Reply:
This is tricky ... the code is generic and using macros which makes it hard to check for literal zero ... I'm going to have to think about this some more. I probably have to refactor this a bit.

Reply:
We could use something like [this](https://rust-num.github.io/num/num/trait.Zero.html).

In [ARROW-3878](https://github.com/apache/arrow/pull/3031) @sunchao is introducing the following:
`pub trait ArrowNumericType: ArrowPrimitiveType {}`

We could make this:
`pub trait ArrowNumericType: ArrowPrimitiveType + Zero {}` and you could make the above generic over `ArrowNumericType`.

I think it's fine to merge as is and open a JIRA for the follow up work, as once ARROW-3878 is merged the solution will change.

Reply:
Thanks @paddyhoran  ... that num crate made things much nicer and now we no longer panic on divide by zero. Glad to see `ArrowNumericType` coming along ... I actually tried to implement that myself this week and failed.